tests/datasets/storages/processors/test_replaced_groups.py

```diff
@@ -198,3 +231,155 @@ def test_query_overlaps_replacements_processor(
     enforcer._set_query_final(query_with_future_timestamp, True)
     enforcer.process_query(query_with_future_timestamp, HTTPRequestSettings())
     assert not query_with_future_timestamp.get_from_clause().final
+
+
+def test_query_has_group_id(query_with_single_group_id: ClickhouseQuery,) -> None:
+    enforcer = PostReplacementConsistencyEnforcer("project_id", ReplacerState.EVENTS)
+
+    set_project_exclude_groups(
+        2,
+        [105, 106, 107],
+        ReplacerState.EVENTS,
+        ReplacementType.EXCLUDE_GROUPS,  # Arbitrary replacement type, no impact on tests
+    )
+
+    # Query looking for groups with no replacements, but project has replacements
+    enforcer._set_query_final(query_with_single_group_id, True)
+    state.set_config("max_group_ids_exclude", 2)
+
+    enforcer.process_query(query_with_single_group_id, HTTPRequestSettings())
+    assert query_with_single_group_id.get_condition() == build_and(
+        build_in("project_id", [2]), build_in("group_id", [101])
+    )
+    assert not query_with_single_group_id.get_from_clause().final
+
+    set_project_exclude_groups(
+        2,
+        [100, 101, 102],
+        ReplacerState.EVENTS,
+        ReplacementType.EXCLUDE_GROUPS,  # Arbitrary replacement type, no impact on tests
+    )
+
+    # Too many groups to exclude, but query only needs one of them
```

Comment:
I would split each of these cases into a separate test. That way you don't have to keep resetting the query object.